git-branch-linearity.sh

```diff
@@ -6,13 +6,26 @@ TARGET_BRANCH="${1:-main}"
 echo "Target branch: $TARGET_BRANCH"
 git fetch origin $TARGET_BRANCH 2> /dev/null
 
-out=$(git log origin/${TARGET_BRANCH}..HEAD --merges --oneline)
+
+# Detection of (PR) merge commits at tip of HEAD:
+target_sha=$(git rev-parse origin/${TARGET_BRANCH})
+# Get SHA of parent commit
+merge_target=$(git log --merges --oneline --parents --no-abbrev-commit HEAD^..HEAD | cut -d" " -f2)
```

Comment:
I understand that if the last commit is a merge commit and has the head of the target branch as the left child, you will ignore it (consider it's the GitHub generated one) and check for merge commit in the below commits.

It's not clear to me if this approach works if someone did the same thing in his/her branch that GitHub CI does, i.e. he/she merged the main branch into its development branch, which we also wouldn't want in case of linearity check.

I tried locally and the check worked, but it was because the origin/main commit was the right child of the merge commit and not the left child.

Are you sure testing always the left child will achieve the expected results?
Unfortunately, I was not able to test it in a real PR, I had some difficulties reproducing the issue.


BTW I thought initially to use something more GitHub specific, using for instance the environment variable exported.
The `GITHUB_SHA` contains the tip commit (which will be the merge commit if one was created)
and `GITHUB_HEAD_REF` contains the source branch name (and is only defined in PR context)

If the 2 don't point to the same thing, we are probably in a GitHub merge commit situation.
Or simpler, if `GITHUB_HEAD_REF` is non-empty, we do `git rev-parse ${GITHUB_HEAD_REF}` to get the tip commit to use.

BTW I found this page informative about the process: https://www.kenmuse.com/blog/the-many-shas-of-a-github-pull-request/

Reply:
Ok I think it will work because locally the left child is the target of the merge (ours) and the right side what is merged into the target (theirs).
A local merge commit will merge main into the current branch, so it will end up on the right.
Whereas GitHub create a new branch from the base commit and merge the pull request branch into it so the origin/main commit will be on the left.

I still think using `GITHUB_HEAD_REF` would be simpler but as you want :-)